### PR TITLE
[ffigen] Migrate to ARC

### PR DESCRIPTION
**Pull request by liamappelbe**
Migrate package:objective_c, ffigen's generated block trampolines, and all the objective_c tests to ARC.

The generated block trampolines have changed from
```Objective-C
typedef void  (^ListenerBlock1)(DummyObject* );
ListenerBlock1 wrapListenerBlock_ObjCBlock_ffiVoid_DummyObject(ListenerBlock1 block) {
  ListenerBlock1 wrapper = [^void(DummyObject* arg0) {
    block([arg0 retain]);
  } copy];
  [block release];
  return wrapper;
}
```
to
```Objective-C
typedef void  (^ListenerBlock1)(DummyObject* );
ListenerBlock1 wrapListenerBlock_ObjCBlock_ffiVoid_DummyObject(ListenerBlock1 block) NS_RETURNS_RETAINED {
  return ^void(DummyObject* arg0) {
    block(objc_retain(arg0));
  };
}
```

We have to retain the args using `objc_retain`/`objc_retainBlock` instead of sending a retain message. Also, instead of explicitly copying the wrapper block, we can just add a `NS_RETURNS_RETAINED` annotation and rely on ARC to copy it for us. Also, instead of releasing the reference to the incoming block in here, we're doing it in Dart land instead.

Other changes:
- Added a compile time check to the trampolines file to make sure it's compiled with ARC.
- Switched from retaining blocks using copy messages and `_Block_copy` to `objc_retainBlock`, which is more stable. Also switched from releasing blocks using `_Block_release` to the ordinary `objc_release` used for all objects.
- All the ffigen tests have been migrated to ARC except for the old automated_ref_count_test, which has been split into an ARC and non-ARC version (arc_test and ref_count_test). The old name was a bit of a misnomer.
- Added a github workflow that verifies that the package:objective_c example builds, to make sure the podspecs are correct (the tests build the package test/setup.dart, so the flutter build path was untested).

Fixes https://github.com/dart-lang/native/issues/1421
Follow up work: https://github.com/dart-lang/native/issues/1446

### CODE REVIEW

**Review by dcharkes — APPROVED** [review]
> Also, instead of releasing the reference to the incoming block in here, we're doing it in Dart land instead.

What is the effect of this?

> * Added a github workflow that verifies that the package:objective_c example builds, to make sure the podspecs are correct (the tests build the package test/setup.dart, so the flutter build path was untested).

Probably leave a TODO there for switching to native assets, as then we don't use podspecs anymore.

> Fixes #1421
> Follow up work: #1446

Do we end up in kind of a half-state here? E.g. we don't support non ARC anymore, but for ARC we ignore a bunch of things which would lead to memory errors. Should #1446 be folded in to this PR? Or is the current PR usable? If so in what cases? And how do users know?

(Side note: Should we have some way of specifying in our workflow that something should be addressed before the next version is released? Maybe we need a release-blocker issue label? Or, is the [tracker](https://github.com/orgs/dart-lang/projects/87) of when we want to release the next version enough? As long as you're the person mostly working on FFIgen, I guess it's fine to leave FFIgen in a unpublishable state on the main branch, but it does prevent us from addressing other bugs and doing a release in the meantime.)

The code changes in this PR, LGTM.
So, LGTM once above question is answered.

**Comment by kekland** [comment]
Somehow after generating code I got one wrapper function that was still using the `retain` keyword:

```objc
#include <stdint.h>
#import <CoreMIDI/CoreMIDI.h>
#import <CloudKit/CloudKit.h>
#import <CoreFoundation/CoreFoundation.h>
#import <AVFoundation/AVFoundation.h>
#import <AVFoundation/AVAssetExportSession.h>
#import <AVFoundation/AVAssetTrackSegment.h>
#import <AVFoundation/AVAssetImageGenerator.h>
#import <AVFoundation/AVMediaFormat.h>
#import <AVFoundation/AVVideoComposition.h>
#import <CoreGraphics/CGBase.h>
#import <CoreGraphics/CGGeometry.h>
#import <CoreGraphics/CGImage.h>
#import <CoreMedia/CMBase.h>
#import <CoreMedia/CMTime.h>
#import <CoreMedia/CMTimeRange.h>
#import <QuartzCore/CADisplayLink.h>

#if !__has_feature(objc_arc)
#error "This file must be compiled with ARC enabled"
#endif

id objc_retain(id);
id objc_retainBlock(id);

typedef void  (^ListenerBlock)(NSDictionary* , struct _NSRange , BOOL * );
ListenerBlock wrapListenerBlock_ObjCBlock_ffiVoid_NSDictionary_NSRange_bool(ListenerBlock block) NS_RETURNS_RETAINED {
  return ^void(NSDictionary* arg0, struct _NSRange arg1, BOOL * arg2) {
    block(objc_retain(arg0), arg1, arg2);
  };
}

//
// The block below has the issue
//

typedef void  (^ListenerBlock1)(id , struct _NSRange , BOOL * );
ListenerBlock1 wrapListenerBlock_ObjCBlock_ffiVoid_objcObjCObject_NSRange_bool(ListenerBlock1 block) NS_RETURNS_RETAINED {
  return ^void(id arg0, struct _NSRange arg1, BOOL * arg2) {
    block([arg0 retain], arg1, arg2);  // <--- here
  };
}

typedef void  (^ListenerBlock2)(NSDate* , BOOL , BOOL * );
ListenerBlock2 wrapListenerBlock_ObjCBlock_ffiVoid_NSDate_bool_bool(ListenerBlock2 block) NS_RETURNS_RETAINED {
  return ^void(NSDate* arg0, BOOL arg1, BOOL * arg2) {
    block(objc_retain(arg0), arg1, arg2);
  };
}

// bunch more below, but everything is fine there
```

Not sure which interface led to the generation of this. If needed, I can share `ffigen.yaml`

**Comment by liamappelbe** [comment]
> Somehow after generating code I got one wrapper function that was still using the `retain` keyword:

Oops. Good catch. Forgot about `id`. Thanks.

**Comment by liamappelbe** [comment]
> > Also, instead of releasing the reference to the incoming block in here, we're doing it in Dart land instead.
> 
> What is the effect of this?

It doesn't make a functional difference, but it means one less ARC hack in the ObjC bindings.

> > * Added a github workflow that verifies that the package:objective_c example builds, to make sure the podspecs are correct (the tests build the package test/setup.dart, so the flutter build path was untested).
> 
> Probably leave a TODO there for switching to native assets, as then we don't use podspecs anymore.

Done.

> > Fixes #1421
> > Follow up work: #1446
> 
> Do we end up in kind of a half-state here? E.g. we don't support non ARC anymore, but for ARC we ignore a bunch of things which would lead to memory errors. Should #1446 be folded in to this PR? Or is the current PR usable? If so in what cases? And how do users know?

This isn't a regression. The migration to ARC just made some existing bugs more obvious. The way we're not quite handling method families correctly is a very obscure issue that users will probably never actually hit. The bigger issue is the way we are assuming all blocks that return ref countable objects are `NS_RETURNS_RETAINED`. That could lead to use after free issues. But in practice, blocks that return objects aren't very common (most blocks are completion handlers that return void), which is why no users have ever reported this issue. But I'll do that follow up in the same 14.0.0 release, so users won't see the half-state anyway.

> Side note: Should we have some way of specifying in our workflow that something should be addressed before the next version is released?

I've been using milestones. Eg, #1446 is in the ffigen 14.0.0 milestone.

File: pkgs/objective_c/CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 - Drop API methods that are deprecated in the oldest versions of iOS and macOS
   that flutter supports.
 - Added `ObjCBlock`, which is the new user-facing representation of ObjC blocks.
+- Migrate to ARC.
```
[the remark this answers is not in the document]
Comment (liamappelbe) [reply pkgs/objective_c/CHANGELOG.md:6]: Done

File: pkgs/objective_c/example/macos/Podfile.lock [hunk 2]
```diff
@@ -0,0 +1,22 @@
+PODS:
```
[the remark this answers is not in the document]
Comment (liamappelbe) [reply pkgs/objective_c/example/macos/Podfile.lock:1]: Removed

**Pull request merged**